File: python_bindings/python/Type.cpp
#include "Type.h"

// to avoid compiler confusion, python.hpp must be include before Halide headers
#include <boost/format.hpp>
#include <boost/python.hpp>

#include "Halide.h"

#include <string>
#include <vector>

namespace h = Halide;

std::string type_code_to_string(const h::Type &t) {
    std::string code_string = "unknown";
    switch (t.code()) {
    case h::Type::UInt:
        code_string = "UInt";
        break;

    case h::Type::Int:
        code_string = "Int";
        break;

    case h::Type::Float:
        code_string = "Float";
        break;

    case h::Type::Handle:
        code_string = "Handle";
        break;

    default:
        code_string = "unknown";
    }

    return code_string;
}

Halide::Type make_handle(int lanes) {
    return Halide::Handle(lanes, nullptr);
}

std::string type_repr(const h::Type &t) {
    auto message_format = boost::format("<halide.Type code '%s' with %i bits and %i lanes>");

    return boost::str(message_format % type_code_to_string(t) % t.bits() % t.lanes());
}

void defineType() {

    using Halide::Type;
    namespace p = boost::python;

    bool (Type::*can_represent_other_type)(Type) const = &Type::can_represent;

    p::class_<Type>("Type",
                    "Default constructor initializes everything to predictable-but-unlikely values",
                    p::no_init)
        .def(p::init<halide_type_code_t, int, int>(p::args("code", "bits", "lanes")))
        .def(p::init<h::Type>(p::args("that"), "Copy constructor"))

        .def("bits", &Type::bits,
             "The number of bits of precision of a single scalar value of this type.")
        .def("bytes", &Type::bytes,
             "The number of bytes required to store a single scalar value of this type. Ignores vector lanes.")
        .def("lanes", &Type::lanes,
             "How many elements (if a vector type). Should be 1 for scalar types.")
        .def("is_bool", &Type::is_bool, p::arg("self"),
             "Is this type boolean (represented as UInt(1))?")
        .def("is_vector", &Type::is_vector, p::arg("self"),
             "Is this type a vector type? (lanes > 1)")
        .def("is_scalar", &Type::is_scalar, p::arg("self"),
             "Is this type a scalar type? (lanes == 1)")
        .def("is_float", &Type::is_float, p::arg("self"),
             "Is this type a floating point type (float or double).")
        .def("is_int", &Type::is_int, p::arg("self"),
             "Is this type a signed integer type?")
        .def("is_uint", &Type::is_uint, p::arg("self"),
             "Is this type an unsigned integer type?")
        .def("is_handle", &Type::is_handle, p::arg("self"),
             "Is this type an opaque handle type (void *)")
        .def(p::self == p::self)
        .def(p::self != p::self)
        .def("with_lanes", &Type::with_lanes, p::args("self", "w"),
             "Produce a copy of this type, with 'lanes' vector lanes")
        .def("with_bits", &Type::with_bits, p::args("self", "w"),
             "Produce a copy of this type, with 'bits' bits")
        .def("element_of", &Type::element_of, p::arg("self"),
             "Produce the type of a single element of this vector type")
        .def("can_represent", can_represent_other_type, p::arg("other"),
             "Can this type represent all values of another type?")
        .def("max", &Type::max, p::arg("self"),
             "Return an expression which is the maximum value of this type")
        .def("min", &Type::min, p::arg("self"),
             "Return an expression which is the minimum value of this type")
        .def("__repr__", &type_repr, p::arg("self"),
             "Return a string containing a printable representation of a Type object.");

    p::def("Int", h::Int,
           (p::arg("bits"), p::arg("lanes") = 1),
           "Constructing an signed integer type");

    p::def("UInt", h::UInt,
           (p::arg("bits"), p::arg("lanes") = 1),
           "Constructing an unsigned integer type");

    p::def("Float", h::Float,
           (p::arg("bits"), p::arg("lanes") = 1),
           "Constructing a floating-point type");

    p::def("Bool", h::Bool,
           (p::arg("lanes") = 1),
           "Construct a boolean type");

    p::def("Handle", make_handle,
           (p::arg("lanes") = 1),
           "Construct a handle type");

    return;
}

File: python_bindings/test/correctness/type.py
import halide as hl


def test_type():
    t1 = hl.Type()
    assert t1.code() == hl.TypeCode.Handle
    assert t1.bits() == 0
    assert t1.lanes() == 0

    t1 = hl.Type(hl.TypeCode.Int, 32, 1)
    t2 = hl.Int(32)
    assert t1 == t2
    assert t2.code() == hl.TypeCode.Int
    assert t2.bits() == 32
    assert t2.lanes() == 1
    assert t2.bytes() == 4

    t1 = t2.with_code(hl.TypeCode.UInt)
    assert t1 == hl.UInt(32)

    t1 = t2.with_bits(16)
    assert t1 == hl.Int(16)
    assert t1 != t2

    t1 = t2.with_lanes(8)
    assert t1 == hl.Int(32, 8)
    assert t1 != t2
    assert t1.element_of() == hl.Int(32)

    b1 = hl.Bool()
    f32 = hl.Float(32)
    h64 = hl.Handle()
    i32 = hl.Int(32)
    u32 = hl.UInt(32)
    vi32x8 = hl.Int(32, 8)
    u64 = hl.UInt(64)
    i64 = hl.Int(64)

    assert b1.is_bool()
    assert f32.is_float()
    assert h64.is_handle()
    assert i32.is_int()
    assert u32.is_uint()
    assert i64.is_int()
    assert u64.is_uint()

    assert not vi32x8.is_scalar()
    assert vi32x8.is_vector()

    h2 = hl.Handle()
    assert h64.same_handle_type(h2)

    assert f32.can_represent(b1)
    assert not b1.can_represent(f32)

    assert not b1.is_max(2)
    assert b1.is_max(1)
    assert b1.is_min(0)
    assert not b1.is_min(-1)

    assert i32.is_max(2147483647)
    assert i32.is_min(-2147483648)

    assert not u32.is_max(4294967296)
    assert u32.is_max(4294967295)
    assert u32.is_min(0)
    assert not u32.is_min(-1)
    assert not u32.is_min(1)

    assert i64.is_max(9223372036854775807)
    assert i64.is_min(-9223372036854775808)

    # Python doesn't have unsigned integers, so we can't really express this value.
    # assert u64.is_max(0xFFFFFFFFFFFFFFFF)
    assert u64.is_min(0)

    # repr() and str()
    assert str(i32) == "int32"
    assert repr(i32) == "<halide.Type int32>"
    assert str(vi32x8) == "int32x8"
    assert repr(vi32x8) == "<halide.Type int32x8>"


if __name__ == "__main__":
    test_type()
